hledger-lib/Hledger/Read/JournalReader.hs

```diff
@@ -552,10 +553,8 @@ defaultyeardirectivep :: JournalParser m ()
 defaultyeardirectivep = do
   char 'Y' <?> "default year"
   lift skipNonNewlineSpaces
-  y <- some digitChar
-  let y' = read y
-  failIfInvalidYear y
-  setYear y'
+  y <- decimal
+  setYear y
```

Comment:
I see you dropped the valid year check. Until now we have required that years are always written with at least 4 digits, to reduce problems parsing ambiguous dates. I guess in a Y directive it doesn't matter, but would the inconsistency be unpleasant ? 4 digits are still needed in most other places, like transaction dates, I imagine.

Reply:
Yes, the only place where the valid year check seemed relevant is to distinguish different incomplete date formats `YYYY-MM` format from `MM-DD` format, and distinguishing "just a year" from "just a day of the month". I've replaced the ‘four digit check’ in those places with `fromGregorianValid`
https://github.com/simonmichael/hledger/blob/96b8761c746ccecf9aba47a8d59862947279aa22/hledger-lib/Hledger/Data/Dates.hs#L799-L802
or checking for a valid day number https://github.com/simonmichael/hledger/blob/96b8761c746ccecf9aba47a8d59862947279aa22/hledger-lib/Hledger/Data/Dates.hs#L832-L837

Looking through the code, I don't think that transaction dates were ever required to be four digits. It looks like they were always checked using `fromGregorianValid`
https://github.com/simonmichael/hledger/blob/96b8761c746ccecf9aba47a8d59862947279aa22/hledger-lib/Hledger/Read/Common.hs#L468-L471

Reply:
Thinking about it more, I agree that we should require at least four digits. People might mistakenly enter a two-digit year thinking that 33 would be completed to 2033 (or 1933). I'll add a proper year parser.

Reply:
Sounds good, it's a simple rule users will remember and hopefully no hardship.

Reply:
Perhaps the datep parser should also be rewritten to do the four-digit check.

Reply:
Year parsing now requires four digits, this includes in the datep parser (which previously didn't check that condition).